### DISCUSSION TITLE
make parameter handling orthogonal, please

### DISCUSSION THREAD

**Discussion by tonne1** [duckdb/duckdb#12505] [General]
### What happens?

For context, please see my discussion #12275 with @soerenwolfers.

Credits for the workaround go to @ggPeti in #11393, it almost went unnoticed.

Orthogonal parameter handling is such a fundamental architectural feature, that this cannot be a feature request, so I made it an issue. I'm aiming at using duckdb as a self-contained incremental data pipeline, as an ETL engine where everything necessary is IN the database or accessible from there. But even if one doesn't go that far down the DB-centric route, duckdb would be an order of magnitude more useful and versatile, if that single issue/showstopper was resolved. Because then, passing intermediate results to succeeding steps would be possible, whether in a query or in a script. Plus of course, the promise of Friendly SQL would be kept.

I'm still fairly new to duckdb, so chances are high that I overlooked a number of aspects. Apologies for that, right now I can only contribute from a user perspective, this is in no way meant to be audacious. Thanks to everyone contributing to this great machine.

### To Reproduce

see #12275

### OS:

any

### DuckDB Version:

1.0.0

### DuckDB Client:

any

### Full Name:

tonne1

### Affiliation:

s8

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Not applicable - the reproduction does not require a data set

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by szarnyasg** [comment]
Thanks @tonne1 for the issue and the explanation, we'll discuss this next week with the team.

**Comment by Mytherin** [comment]
Thanks for reporting!

DuckDB already supports this in almost all instances. The main difference between DuckDB and other database engines (like Postgres) is that DuckDB supports much more flexible table functions. In particular, DuckDB supports table functions whose output schema is **dependent on the input**. For example, DuckDB supports the `read_csv` function:

```sql
SELECT * FROM read_csv('file.csv');
```

The column names and column types that are returned from `read_csv`, are dependent on the **contents** of the file `file.csv`.

Now imagine how that function would work if we were to read the parameters to this function from another part of the query, e.g.:

```sql
SELECT * FROM tbl, read_csv(tbl.file);
```

In order for this to be handled, we would essentially have to read `tbl` in order to figure out which files would be read, and then continue query planning. Of course it can get worse:

```sql
SELECT * FROM (SELECT MAX(file) FROM tbl) subq, read_csv(subq.file);
```

Now in order to process this query, we need to **execute a subquery**, and then continue query planning.

Because this would complicate the system immensely - these types of functions always require literals as inputs. This is very unlikely to change because of the complexity it would bring along with it. I am also not aware of any other system that supports this. Postgres does not support these kinds of functions at all, and systems that do generally do not support lateral joins at all.

That being said - there are functions that are currently unnecessarily restrictive, mostly for legacy reasons. As mentioned in the linked discussion the `range` function is one of those - it was implemented using the same table function API as `read_csv` as it predates lateral join support entirely. There have been some attempts to fix this before but unfortunately they were not successful. While there is a work-around available (`SELECT UNNEST(RANGE(...))`) - this has come up frequently and seems to be a point of confusion for users. I've decided to pick this up myself and have opened a PR that implements this here - https://github.com/duckdb/duckdb/pull/12431.

**Comment by tonne1** [comment]
Thanks @Mytherin for taking this up and for your explanations! I would like to comment on some:

- table functions whose output schema is dependent on their input

Actually those are a fraction of all functions - restricting parameter transport for all others too is a unnecessary high price IMO.
Parameter transport is crucial for control flow, data flow and reusable code.
Of course one wants to avoid unnecessary complexity and performance regression in the query analyzer, I fully agree.
But in most cases you can get the necessary metadata to decide in which context an expression is applicable, at an early stage, when a function gets created, without running the query.

We have dynamic schemas in other areas, e.g. pivot, union (of tables), read_parquet([p1, p2, ...]), ... 
There are some mechanisms already in place to deal with dynamism, union-by-name, filename, positional joins, columns(*), group by all etc. 

Please lets, for a moment, distinguish between input parameter transport and requirements/restrictions imposed by dynamic output schemas.

The following example uses a configuration value to build the path:

```sql
select filename path, last_modified 
from read_blob(static_conf('basedir') || '/parquet/*.parquet');
```

and with this function it runs fine:

```sql
create or replace function static_conf(p) as
case p
  when 'basedir' then '/path/to/base'
  when 'baseurl' then 'https://path/to/base'
  -- 50 more settings
  ...
  else null 
end;
```

But if the mapping of params is in a conf table, were out of luck. Sure we can define

```sql
create or replace function table_conf(p) as table
select val from conf where param = p limit 1;
```

or

```sql
create or replace function table_conf(p) as
(select val from conf where param = p)::varchar;
```

But none of them can be used in the query above ("Table function cannot contain subqueries"), although the result of the functions is the same, a scalar varchar. 
By the way, the output structure of the query is not dependent on the input, it will always be varchar and timestamp.

This means that through some programming API or by scripting, one either has to setup the static macro or generate code which inserts configuration values as literals whereever necessary. 
That is not a good workflow and not reusable, readable code. Considerable overhead for maintenance/testing, the necessity to have a programming API or scripting environment available when setting this up and whereever you deploy to. And this was just a really simple example.

If all that's needed is to signal the analyzer what to expect, I'd be happy to include a flag like 'scalar', additional type information etc, so that the analyzer can easily determine in which context this thing is applicable. 

- how other RDBMSes do this

Other DBMSes do it that way - they request, even enforce as much context from the user as possible, and thats fair enough. 
Let us write clean, concise, reusable code, which can be as complex as the task requires, and we provide all required hints for the analyzer. 
In cases which are fully dynamic re output schemas, one can still impose restrictions.

Postgres asks for a number of additional clues when you define a stored function: https://www.postgresql.org/docs/current/sql-createfunction.html 

Postgres and Firebird use a different terminology re stored functions and procedures. 
In Postgres a procedure is a function which does not return a result set.
In Firebird they distinguish between executable and selectable stored procedures. Selectable stored procedures can be used like a table which can take parameters, while executable procedures are the equivalent of Postgres procedures.
Stored functions in Firebird only ever return a scalar value, but accept 0 - n input params. https://firebirdsql.org/file/documentation/html/en/refdocs/fblangref50/firebird-50-language-reference.html#fblangref50-psql-storedfuncs 

So in Firebird you would do this:

```sql
create or alter function conf (param varchar(100))
returns varchar(300)
as
begin
  return (select val from conf where param = :param);
end;
```

or even this:

```sql
create or alter function conf (param varchar(100))
returns varchar(300)
as
declare val varchar(300)
begin
  select val from conf where param = :param into :val;
  if (:val is null)
    -- request the value from a web server and update the conf table
  return :val;
end;
```

By using a stored function you tell the system it can expect scalar output, and you even give the type. Now the system can tell where this function can be used. 

and then

```sql
select conf('basedir') from rdb$database; -- the equivalent to oracle's 'dual'
select * from split('hello world ' || conf('basedir'), ' ');
select * from split('hello world ' || (select val from conf where param = 'basedir'), ' ');
-- and of course lateral joins
select s.* 
from (select conf('basedir') val from rdb$database) c
join lateral (select * from split(c.val, ' ')) s 
```

In the example split is a (selectable) stored procedure, semantics "split_to_table", but could be any code. The stored function conf can be used whereever a scalar expression is applicable. 
It doesn't matter at all that a stored function creates the parameter.

Language mechanisms for parameter transport in Firebird (across queries and scripts):

- orthogonal parameter handling
- lateral joins
- arbitrary nesting of stored functions and procedures
- stored functions for scalar output
- selectable stored procedures for tabular output
- execute blocks (anonymous procedures, run like query)
- context variables (store/retrieve values withtin transaction/session, in different name spaces)
- execute statement - run a textual sql statement, also on a remote database 
- a procedural language

You might ask, why are you here then? Because DuckDB is optimized for OLAP, supports a lot of data interfaces and has a rich SQL language, and that's what I need. The only missing bit is the ability to fully combine the features DuckDB offers.

**Comment by Mytherin** [comment]
Thanks for expanding - but I still don't see anything that's *fundamentally* missing. DuckDB supports everything that you mention in all contexts *outside* of table functions that are designed to take single constants as parameters for reasons previously stated. As mentioned, that is highly unlikely to change in the general case. There might be a few functions that are unnecessarily restricted by this - like the `range` function was - but we can deal with that on a case-by-case basis.

As such this issue isn't particularly actionable and I'll move this back to a discussion. If you encounter functions that you believe to be unnecessarily restrictive feel free to open separate actionable issues.

**Reply by cmettler** [reply]
I've encountered situations where I needed to dynamically construct parameters for a table macro used in a lateral join. While I've resorted to variables or dbt, I'd prefer a more direct SQL-based solution.

In MS SQL Server, cross/outer apply patterns, similar as described by @tonne1, are often used for this. I'm curious if a similar approach is possible in DuckDB, even if it leads to less efficient nested loops .

**Comment by tonne1** [comment]
Thanks for taking the time, that's reasonable. I'll approach this in smaller steps.

Ist there a way to detect (other than through an error message) which functions expect constants? Through duckdb_functions() ?

**Reply by Tishj** [reply]
Table functions all expect constants, except for table in-out functions as mentioned [here](https://github.com/duckdb/duckdb/discussions/12505#discussioncomment-9754240)

**Reply by cmettler** [reply]
Thanks Tishj
